src/main.rs

```diff
@@ -30,82 +31,126 @@ mod error;
+type Looper = looper::Looper<PortMidiNoteTracker>;
 
-    let (input_id, output_id) = {
-        let args: Vec<String> = std::env::args().collect();
-
-        if args.len() < 2 {
-            print_devices(&context);
-            println!("Usage: ./dimooper <input-port> <output-port>");
-            std::process::exit(1);
-        }
-
-        (args[1].trim().parse().unwrap(), args[2].trim().parse().unwrap())
-    };
-
-    let in_info = context.device(input_id).unwrap();
-    println!("Listening on: {} {}", in_info.id(), in_info.name());
-    let in_port = context.input_port(in_info, 1024).unwrap();
-
-    let out_info = context.device(output_id).unwrap();
+fn create_looper(context: &pm::PortMidi, output_id: DeviceId) -> Result<Looper> { 
+    let out_info = try!(context.device(output_id));
     println!("Sending recorded events: {} {}",
              out_info.id(),
              out_info.name());
-    let out_port = context.output_port(out_info, 1024).unwrap();
+    let out_port = try!(context.output_port(out_info, 1024));
+    let looper = looper::Looper::new(PortMidiNoteTracker::new(out_port));
+    Ok(looper)
+}
 
-    let window_width = RATIO_WIDTH * RATIO_FACTOR;
+fn create_event_loop(context: &pm::PortMidi, input_id: DeviceId) -> Result<EventLoop<'static>> {
+    let window_width = RATIO_WIDTH  * RATIO_FACTOR;
     let window_height = RATIO_HEIGHT * RATIO_FACTOR;
-    let sdl_context = sdl2::init().unwrap();
-    let video_subsystem = sdl_context.video().unwrap();
-    let timer_subsystem = sdl_context.timer().unwrap();
-    let ttf_context = sdl2_ttf::init().unwrap();
 
-    let window = video_subsystem.window("Dimooper", window_width, window_height)
+    let in_info = try!(context.device(input_id));
+    println!("Listening on: {} {}", in_info.id(), in_info.name());
+    let in_port = try!(context.input_port(in_info, 1024));
+
+    let sdl_context = try!(sdl2::init());
+    let video_subsystem = try!(sdl_context.video());
+    let timer_subsystem = try!(sdl_context.timer());
+
+    let window = try!(video_subsystem.window("Dimooper", window_width, window_height)
         .position_centered()
         .opengl()
-        .build()
-        .unwrap();
-
-    let renderer = window.renderer().build().unwrap();
+        .build());
 
-    let bpm_popup = {
-        let font = ttf_context.load_font(Path::new(TTF_FONT_PATH), 50).unwrap();
-        Popup::new(font)
-    };
+    let renderer = try!(window.renderer().build());
+    let event_pump = try!(sdl_context.event_pump());
+    let event_loop = EventLoop::new(timer_subsystem, event_pump, in_port, renderer);
+    Ok(event_loop)
+}
 
-    let event_pump = sdl_context.event_pump().unwrap();
+fn init_font() -> Result<Popup> {
+    let ttf_context = try!(sdl2_ttf::init());
+    let font = try!(ttf_context.load_font(Path::new(TTF_FONT_PATH), 50));
+    let popup = Popup::new(font);
+    Ok(popup)
+}
 
-    let looper = looper::Looper::new(PortMidiNoteTracker::new(out_port));
+fn main() {
```

Comment:
It can be done with easily `SubcommandRequiredElseHelp` or `ArgsRequiredElseHelp` settings (see https://docs.rs/clap/2.25.1/clap/enum.AppSettings.html), but it contradicts with your view of how `help` subcommand should work. Alternatively, we can handle case when no arguments provided manually. 

IMO, program should not show list of devices by default. It causes 'WTF' effect. I think it should show what that program is (about) and how to use it (help).